sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,334 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of the {@link CassandraService} that actually use a Cassandra instance.
+ */
+public class CassandraServiceImpl<T> implements CassandraService<T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CassandraServiceImpl.class);
+
+  private class CassandraReaderImpl<T> extends BoundedSource.BoundedReader<T> {
+
+    private final CassandraIO.CassandraSource<T> source;
+
+    private Cluster cluster;
+    private Session session;
+    private ResultSet resultSet;
+    private Iterator<T> iterator;
+    private T current;
+
+    public CassandraReaderImpl(CassandraIO.CassandraSource<T> source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOG.debug("Starting Cassandra reader");
+      cluster = getCluster(source.spec.hosts(), source.spec.port());
+      session = cluster.connect();
+      LOG.debug("Query: " + source.splitQuery);
+      resultSet = session.execute(source.splitQuery);
+
+      final MappingManager mappingManager = new MappingManager(session);
+      Mapper mapper = mappingManager.mapper(source.spec.entity());
+      iterator = mapper.map(resultSet).iterator();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (iterator.hasNext()) {
+        current = iterator.next();
+        return true;
+      }
+      current = null;
+      return false;
+    }
+
+    @Override
+    public void close() {
+      LOG.debug("Closing Cassandra reader");
+      if (session != null) {
+        session.close();
+      }
+      if (cluster != null) {
+        cluster.close();
+      }
+    }
+
+    @Override
+    public T getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public CassandraIO.CassandraSource<T> getCurrentSource() {
+      return source;
+    }
+
+  }
+
+  @Override
+  public CassandraReaderImpl<T> createReader(CassandraIO.CassandraSource<T> source) {
+    return new CassandraReaderImpl<>(source);
+  }
+
+  @Override
+  public long getEstimatedSizeBytes(CassandraIO.Read<T> spec) {
+    try (Cluster cluster = getCluster(spec.hosts(), spec.port())) {
+      List<TokenRange> tokenRanges = getTokenRanges(cluster,
+          spec.keyspace(),
+          spec.table());
+      return getEstimatedSizeBytes(cluster, tokenRanges);
+    }
+  }
+
+  /**
+   * Actually estimate the size of the data to read on the cluster, based on the given token
+   * ranges to address.
+   */
+  @VisibleForTesting
+  protected static long getEstimatedSizeBytes(Cluster cluster, List<TokenRange> tokenRanges) {
+    long size = 0L;
+    for (TokenRange tokenRange : tokenRanges) {
+      size += tokenRange.meanPartitionSize * tokenRange.partitionCount;
+    }
+    if (getRingFraction(cluster, tokenRanges).compareTo(new BigDecimal("0")) == 1) {
```

Comment:
If the token ranges is very small compared to the global cluster, then ring fraction (due to the rounding) is 0.